core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1936,7 +1984,311 @@ protected Scope visitUpdate(Update update, Optional<Scope> scope)
+                allColumnNamesBuilder.addAll(mergeColumnNames);
+                mergeCaseColumnsListsBuilder.put(caseCounter, mergeColumnNames);
+                if (operation instanceof MergeInsert) {
+                    caseDetailsBuilder.add(new MergeCaseDetails(
+                            caseCounter,
+                            MergeCaseKind.INSERT,
+                            ImmutableSet.copyOf(mergeColumnNames)));
+                }
+                else if (operation instanceof MergeUpdate) {
+                    caseDetailsBuilder.add(new MergeCaseDetails(caseCounter, MergeCaseKind.UPDATE, ImmutableSet.copyOf(mergeColumnNames)));
+                }
+                else if (operation instanceof MergeDelete) {
+                    caseDetailsBuilder.add(new MergeCaseDetails(caseCounter, MergeCaseKind.DELETE, ImmutableSet.of()));
+                }
+                else {
+                    throw new IllegalArgumentException(format("Unknown MergeOperation %s of class %s", operation, operation.getClass().getName()));
+                }
+                caseCounter++;
+            }
+            List<MergeCaseDetails> mergeCases = caseDetailsBuilder.build();
+            Map<Integer, List<String>> mergeCaseColumnsLists = mergeCaseColumnsListsBuilder.build();
+            Set<String> allColumnsUpdated = allColumnNamesBuilder.build();
+            MergeDetails mergeDetails = new MergeDetails(mergeCases);
+
+            // Build the required mappings between table column name and column type
+            List<String> allColumnNames = allColumns.stream().map(ColumnMetadata::getName).collect(toImmutableList());
+
+            List<NameAndType> columnTypes = allColumns.stream()
+                    .map(metadata -> new NameAndType(metadata.getName(), metadata.getType()))
+                    .collect(toImmutableList());
+            Map<String, Type> allUpdatedColumnTypes = createColumnTypeMap(columnTypes, allColumnsUpdated);
+            Map<String, Type> allColumnTypes = createColumnTypeMap(columnTypes, allColumnNames);
+
+            // Create the RowType that holds all column values
+            ImmutableList.Builder<RowType.Field> fieldsBuilder = ImmutableList.builder();
+            allColumnTypes.forEach((columnName, type) -> fieldsBuilder.add(new RowType.Field(Optional.of(columnName), type)));
+            // Add the case number and the operation number
+            fieldsBuilder.add(new RowType.Field(Optional.empty(), INTEGER));
+            fieldsBuilder.add(new RowType.Field(Optional.empty(), INTEGER));
+            List<RowType.Field> updatedColumnFields = fieldsBuilder.build();
+            RowType rowType = RowType.from(updatedColumnFields);
+
+            // Perform legality and access control checks
+            accessControl.checkCanMergeTableColumns(session.toSecurityContext(), tableName, allColumnsUpdated);
```

Comment:
We don't need a new access control method. Instead, we can check for insert, update, and delete separately as relevant, which is how the SQL standard implements it.

Reply:
Excellent point.  I removed the merge permissions machinery, and instead iterate over the `MergeDetails` cases, and for the INSERT cases, I check insert permissions; for the DELETE cases, I check delete permissions, and for the UPDATE cases I check update permissions on the columns updated in that case.